lib/src/signing.rs

```diff
@@ -0,0 +1,217 @@
+    /// backend is not configured.
+    #[default]
+    Drop,
+    /// Only sign commits that were authored by self and already signed,
+    /// "preserving" the signature across rewrites.
+    /// This is what jj does when a signing backend is configured.
+    Keep,
+    /// Sign/re-sign commits that were authored by self. Foreign signatures are
+    /// overwritten on authored commits and dropped on others.
+    /// This is what jj does when configured to always sign.
+    Own,
+    /// Always sign commits, regardless of who authored or signed them before.
+    /// This is what jj does on `jj sign -f` or `jj --force-sign <snapshotting
+    /// command>`.
+    Force,
+}
+
+impl SignBehavior {
+    pub fn from_config(sign_all: bool) -> Self {
+        match sign_all {
+            true => SignBehavior::Own,
+            false => SignBehavior::Keep,
+        }
+    }
+}
+
+/// Wraps low-level signing backends and adds caching, similar to `Store`.
+#[derive(Debug, Default)]
+pub struct Signer {
+    main_backend: Option<Box<dyn SigningBackend>>,
+    other_backends: Vec<Box<dyn SigningBackend>>,
+    cache: RwLock<HashMap<CommitId, Verification>>,
+}
+
+impl Signer {
+    pub fn from_settings(settings: &UserSettings) -> Self {
+        let mut other_backends: Vec<Box<dyn SigningBackend>> = vec![
+            // todo implement some backends
+        ];
+
+        let main_backend = settings
+            .signing_backend()
+            .and_then(|backend| other_backends.iter().position(|b| b.name() == backend))
+            .map(|pos| other_backends.swap_remove(pos));
+
+        Self::new(main_backend, other_backends)
+    }
+
+    pub fn new(
+        main_backend: Option<Box<dyn SigningBackend>>,
+        other_backends: Vec<Box<dyn SigningBackend>>,
+    ) -> Self {
+        Self {
+            main_backend,
+            other_backends,
+            cache: Default::default(),
+        }
+    }
+
+    pub fn can_sign(&self) -> bool {
+        self.main_backend.is_some()
+    }
+
+    /// This is just a pass-through to the main backend that unconditionally
+    /// creates a signature.
+    pub fn sign(&self, data: &[u8], key: Option<&str>) -> Result<Vec<u8>, SignError> {
+        self.main_backend
+            .as_ref()
+            .expect("tried to sign without checking can_sign first")
+            .sign(data, key)
+    }
+
+    /// Looks for backend that can verify the signature and returns the result
+    /// of its verification.
+    pub fn verify(&self, commit_id: &CommitId, data: &[u8], signature: &[u8]) -> Verification {
+        let cached = self.cache.read().unwrap().get(commit_id).cloned();
+        if let Some(check) = cached {
+            return check;
+        }
+        if let Some(backend) = self
+            .main_backend
+            .iter()
+            .chain(self.other_backends.iter())
+            .find(|b| b.is_of_this_type(signature))
+        {
+            let Ok(check) = backend.verify(data, signature) else {
```

Comment:
nit: I think the backend error would have to be propagated to the caller.

(Error handling can be added later.)